File: api-java/src/main/java/fr/huiitre/tools/modules/riot/sync/application/usecase/SyncValorantUseCase.java
package fr.huiitre.tools.modules.riot.sync.application.usecase;

import java.util.Optional;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.sync.application.ValorantGlobalSyncReport;
import fr.huiitre.tools.modules.riot.sync.application.ValorantSyncReport;
import fr.huiitre.tools.modules.riot.sync.application.ValorantWeaponSyncResult;

@Service
@Transactional
public class SyncValorantUseCase implements SecuredUseCase {

    private final SyncValorantContentTiersUseCase syncContentTiersUseCase;
    private final SyncValorantWeaponsUseCase syncWeaponsUseCase;
    private final SyncValorantSkinsUseCase syncSkinsUseCase;
    private final SyncValorantBundlesUseCase syncBundlesUseCase;

    public SyncValorantUseCase(
            SyncValorantContentTiersUseCase syncContentTiersUseCase,
            SyncValorantWeaponsUseCase syncWeaponsUseCase,
            SyncValorantSkinsUseCase syncSkinsUseCase,
            SyncValorantBundlesUseCase syncBundlesUseCase) {
        this.syncContentTiersUseCase = syncContentTiersUseCase;
        this.syncWeaponsUseCase = syncWeaponsUseCase;
        this.syncSkinsUseCase = syncSkinsUseCase;
        this.syncBundlesUseCase = syncBundlesUseCase;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.TECH;
    }

    public ValorantGlobalSyncReport execute() {
        ValorantSyncReport contentTiers = syncContentTiersUseCase.execute();
        ValorantWeaponSyncResult weaponsResult = syncWeaponsUseCase.execute();
        ValorantSyncReport skins = syncSkinsUseCase.execute(weaponsResult.weaponAssetIdToDbId());
        ValorantSyncReport bundles = syncBundlesUseCase.execute();
        return new ValorantGlobalSyncReport(contentTiers, weaponsResult.report(), skins, bundles);
    }
}

File: api/Modules/Riot/Sync/Application/Usecases/SyncValorantUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Ports;

namespace Tools.Api.Modules.Riot.Sync.Application.Usecases;

// Synchronisation complète du catalogue Valorant depuis les assets locaux.
// L'ordre n'est pas négociable : les armes avant les skins, qui s'y rattachent, et les raretés
// avant tout le reste. Le tout dans une seule transaction — une passe interrompue au milieu
// laisserait des skins orphelins de leur arme.
public sealed class SyncValorantUseCase(
    SyncValorantContentTiersUseCase syncContentTiersUseCase,
    SyncValorantWeaponsUseCase syncWeaponsUseCase,
    SyncValorantSkinsUseCase syncSkinsUseCase,
    SyncValorantBundlesUseCase syncBundlesUseCase,
    ITransactionManager transactionManager
)
{
    public async Task<ValorantGlobalSyncReport> Execute()
    {
        await using var transaction = await transactionManager.BeginAsync();

        var contentTiers = await syncContentTiersUseCase.Execute();
        var weapons = await syncWeaponsUseCase.Execute();
        var skins = await syncSkinsUseCase.Execute(weapons.WeaponAssetIdToDbId);
        var bundles = await syncBundlesUseCase.Execute();

        await transaction.CommitAsync();

        return new ValorantGlobalSyncReport(contentTiers, weapons.Report, skins, bundles);
    }
}
